firedrake/slate/slate.py

```diff
@@ -0,0 +1,616 @@
+
+    prec = None
+    op = "Inverse"
+
+    @classmethod
+    def check_operand_inversion(cls, A):
+        """Ensures that the tensor satisfies the necessary conditions for taking
+        inverses.
+        """
+        assert A.rank == 2, "The tensor must be rank 2."
+        assert A.shape[0] == A.shape[1], "The inverse can only be computed on square tensors."
+
+    @classmethod
+    def get_uop_arguments(cls, A):
+        """Returns the expected arguments of the resulting tensor of
+        performing a specific unary operation on a tensor.
+        """
+        return A.arguments()[::-1]
+
+
+class Transpose(UnaryOp):
+    """An abstract SLATE class representing the tranpose of a tensor."""
+
+    prec = None
+    op = "Transpose"
+
+    @classmethod
+    def get_uop_arguments(cls, A):
+        """Returns the expected arguments of the resulting tensor of
+        performing a specific unary operation on a tensor.
+        """
+        return A.arguments()[::-1]
+
+
+class Negative(UnaryOp):
+    """Abstract SLATE class representing the negation of a tensor object."""
+
+    prec = 1
+    op = "Negative"
+
+    @classmethod
+    def get_uop_arguments(cls, A):
+        """Returns the expected arguments of the resulting tensor of
+        performing a specific unary operation on a tensor.
+        """
+        return A.arguments()
+
+
+class BinaryOp(TensorBase):
+    """An abstract SLATE class representing binary operations on tensors.
+    Such operations take two operands and returns a tensor-valued expression.
+
+    The currently supported binary operations include the following:
+
+        (1) The addition of two identical-rank :class:`TensorBase` objects. That is,
+            the addition of two scalars, vectors or matrices. This operation is
+            implemented in the subclass `TensorAdd`.
+        (2) The subtraction of two scalar, vector or matrix expressions. See `TensorSub`.
+        (3) The multiplication of two matrices, the action of a matrix on a vector or
+            the action of a scalar on any other rank tensor. All cases are handled by the
+            `TensorMul` subclass.
+
+    :arg A: a :class:`TensorBase` object. This can be a terminal tensor object
+            (:class:`Tensor`) or any derived expression resulting from any number
+            of linear algebra operations on `Tensor` objects. For example,
+            another instance of a `BinaryOp` object is an acceptable input, or
+            a `UnaryOp` object.
+    :arg B: a :class:`TensorBase` object.
+    """
+
+    def __init__(self, A, B):
+        """Constructor for the BinaryOp class."""
+        assert isinstance(A, TensorBase) and isinstance(B, TensorBase), ("Both operands must be slate.TensorBase objects. The operands given are of type (%s, %s)" % (type(A), type(B)))
+        self.check_dimensions(A, B)
+        self.tensors = A, B
+        super(BinaryOp, self).__init__()
```

Comment:
Can we have another try at moving the call of the super class constructor before the attribute assignments on `self`?